src/js/index.d.ts

```diff
@@ -0,0 +1,86 @@
+/**
+ * A list of routes and their parameters and bindings.
+ *
+ * Extended and filled by the route list generated automatically by Ziggy.
+ */
+export interface RouteList {
+    //
+}
+
+/**
+ * A route name registered with Ziggy.
+ */
+type KnownRouteName = keyof RouteList;
+
+/**
+ * A route name, or any string.
+ */
+type RouteName = KnownRouteName | (string & {});
+
+/**
+ * Information about a single route parameter.
+ */
+type ParameterInfo = { name: string, binding?: string };
+
+/**
+ * A primitive route parameter value.
+ */
+type ParameterValue = string | number;
+
+/**
+ * A route parameter value, possibly nested inside an object under its binding key.
+ */
+type ValueOrBoundValue<I extends ParameterInfo> = I extends { binding: string }
+    ? { [K in I['binding']]: ParameterValue } | ParameterValue
+    : ParameterValue
+// type A = ValueOrBoundValue<{ name: 'foo', binding: 'bar' }>;
+// = ParameterValue | { bar: ParameterValue }
+
+/**
+ * An object containing a special '_query' key to target the query string of a URL.
+ */
+type HasQueryParam = { _query?: Record<string, any> };
+/**
+ * An object of parameters for an unknown route.
+ */
+type GenericRouteParamsObject = Record<keyof any, any> & HasQueryParam;
+/**
+ * An object of parameters for a specific named route registered with Ziggy.
+ */
+type KnownRouteParamsObject<I extends readonly ParameterInfo[]> = { [T in I[number] as T['name']]?: ValueOrBoundValue<T> | string } & HasQueryParam;
+/**
+ * An object of route parameters.
+ */
+type RouteParamsObject<N extends RouteName> = N extends KnownRouteName ? KnownRouteParamsObject<RouteList[N]> : GenericRouteParamsObject;
+
+/**
+ * An array of parameters for an unknown route.
+ */
+type GenericRouteParamsArray = ValueOrBoundValue<ParameterInfo>[];
+/**
+ * An array of parameters for a specific named route registered with Ziggy.
+ */
+type KnownRouteParamsArray<I extends readonly ParameterInfo[]> = { [K in keyof I]: ValueOrBoundValue<I[K]> | string };
```

Comment:
How does this type become an array? I can see that it works, but I don't understand how—the type definition `{ [K in keyof I]: ValueOrBoundValue<I[K]> | string }` looks like an object type, I would have expected it to produce something like `{ uuid: ParameterValue }`. Does it have to do with the `keyof` in `K in keyof I`?

Same question here about `| string` too, seems to work fine without it?

Reply:
> How does this type become an array? I can see that it works, but I don't understand how—the type definition { [K in keyof I]: ValueOrBoundValue<I[K]> | string } looks like an object type, I would have expected it to produce something like { uuid: ParameterValue }. Does it have to do with the keyof in K in keyof I?

That was difficult for me as well: Basically, since `I` is an array, `keyof I` is a number. On a `readonly` array, this expression allows to remap the type in an const array. I did not find something about this behavior in the TS documentation, but only on SO and TS seems to be aware of this particular expression and then makes the result an array type again. (What looks to you as an object like `{ 0: MyType, 1: MyType }` is considered as array `[MyType, MyType]`

Caution: Be aware, this expression is very sensible. E.g. the result becomes an object if you directly make a union with this expression like `{ [K in keyof I]: SomType } | string`.

Additionally, If I remember correctly, this is tied to one of the more recent versions of TypeScript.

Reply:
Wild. Thanks so much!